model/flow/protocol_state.go

```diff
@@ -0,0 +1,203 @@
+package flow
+
+import "fmt"
+
+// DynamicIdentityEntry encapsulates nodeID and dynamic portion of identity.
+type DynamicIdentityEntry struct {
+	NodeID  Identifier
+	Dynamic DynamicIdentity
+}
+
+type DynamicIdentityEntryList []*DynamicIdentityEntry
+
+// ProtocolStateEntry holds information about the protocol state at some point in time.
+// It allows to reconstruct the state of identity table using epoch setup events and dynamic identities.
+// It tracks attempts of invalid state transitions.
+// It also holds information about the next epoch, if it has been already committed.
+// This structure is used to persist protocol state in the database.
+type ProtocolStateEntry struct {
+	// setup and commit event IDs for current epoch.
+	CurrentEpochEventIDs EventIDs
+	// setup and commit event IDs for previous epoch.
+	PreviousEpochEventIDs EventIDs
+	// Part of identity table that can be changed during the epoch.
+	// Always sorted in canonical order.
+	Identities DynamicIdentityEntryList
+	// InvalidStateTransitionAttempted encodes whether an invalid state transition
+	// has been detected in this fork. When this happens, epoch fallback is triggered
+	// AFTER the fork is finalized.
+	InvalidStateTransitionAttempted bool
+	// NextEpochProtocolState describes protocol state of the next epoch
+	NextEpochProtocolState *ProtocolStateEntry
```

Comment:
`NextEpochProtocolState` also contains a lot of duplicated data from `ProtocolStateEntry`. What if we replaced the field `NextEpochProtocolState` with:
- a field, `NextEpochEventIDs`
- a method, `NextEpochProtocolState()`, which aggregates the next epoch protocol state

I feel this would simplify the logic associated with this data structure.

Reply:
added as TODO to issue https://github.com/onflow/flow-go/issues/4649

Reply:
Some more thoughts on [Jordan's first point](https://github.com/onflow/flow-go/pull/4559#discussion_r1296532771) (mostly for my own documentation):
* I agree with your suggestion to introduce a state machine that processes IdentityTable-changing events and outputs a new `IdentityTable` 
* Nevertheless, I think the current approach has its own benefits:
   * For consensus, we explicitly accept QC and TC contributions of any consensus participant that was listed in the Epoch setup event. If a malicious node is ejected from the protocol, consensus replicas will drop its votes and timeout messages. Nevertheless, they would _not_ reject a QC or TC if another colluding node included the malicious node's vote or timeout message.

Reply:
Regarding your [earlier comment](https://github.com/onflow/flow-go/pull/4559#discussion_r1298905624) @jordanschalm 
> `NextEpochProtocolState` also contains a lot of duplicated data from `ProtocolStateEntry`

As I have explained in detail in [[Proposal] Detailed specification for framework functionality to support future slashing](https://www.notion.so/dapperlabs/Proposal-Detailed-specification-for-framework-functionality-to-support-future-slashing-37b7d127c75d4834becdf3d38cb2731d?pvs=4), I believe is is fundamentally necessary to track the identity table for the next and the current epoch separately, which in turn necessitates data duplication. 

I share your desire to simplify this. Unfortunately, I think this will result in a computational framework that is too weak to satisfy the requirements. 

Suggestion:
* I don't think adding the sanity check to _this_ version of the code is harmful. If we don't change it, the sanity check makes sense and is helpful. 
* If we change the code, we will be anyway refactoring this file substantially and can through out the sanity check. It won't be any significant work.

Reply:
Outstanding issues are captured in https://github.com/onflow/flow-go/issues/4649 to address later. Resolving...